File: old/spoj/cpp/inversion-count.cpp
/**
 * https://www.spoj.com/problems/INVCNT/
 **/
#include<bits/stdc++.h>
#define lli long long int
#define vi vector<int>
using namespace std;

class MergeSorter {
    private:
        lli num_inversion;

        void reset() {
            num_inversion = 0;
        }

        void merge(vi &list, int start, int mid, int end) {
            int i=start;
            int j=mid+1;
            int k=0;
            vi mergedList(end-start+1);
            for(; i<=mid && j <= end;) {
                if (list[i] <= list[j]) {
                    mergedList[k++] = list[i++];
                } else {
                    num_inversion += (mid-i+1);
                    mergedList[k++] = list[j++];
                }
            }

            while(i <= mid) {
                mergedList[k++]=list[i++];
            }

            while(j<=end) {
                mergedList[k++]=list[j++];
            }

            for(int k=0; k<mergedList.size(); k++) {
                list[start+k] = mergedList[k];
            }
        }

        void do_merge_sort(vi &list, int start, int end) {
            if (start < end) {
                int mid = start + (end - start) / 2;
                do_merge_sort(list, start, mid);
                do_merge_sort(list, mid+1, end);
                merge(list, start, mid, end);
            }
        }
    public:
        void merge_sort(vi &list) {
            reset();
            do_merge_sort(list, 0, list.size()-1);
        }

        lli get_num_inversions() {
            return this->num_inversion;
        }
};

int main() {
    MergeSorter merge_sorter;
    int n;
    int t;
    cin>>t;
    while(t--) {
        cin >> n;
        vi list(n);
        for(int i=0; i<n; i++)
            cin>>list[i];
        merge_sorter.merge_sort(list);
        cout<<merge_sorter.get_num_inversions()<<"\n";
    }
}

File: old/spoj/golang/inversion_count.go
/**
 * https://www.spoj.com/problems/INVCNT/
 **/
package main

type mergeSorterInversionCount struct {
	invCount int64
}

func (m *mergeSorterInversionCount) mergeInversionCount(list []int, s, mid, e int) {
	i, j, k := s, mid+1, 0
	mergedList := make([]int, e-s+1)

	for i <= mid && j <= e {
		if list[i] <= list[j] {
			mergedList[k] = list[i]
			k++
			i++
		} else {
			m.invCount += (int64)(mid - i + 1)
			mergedList[k] = list[j]
			k++
			j++
		}
	}

	for i <= mid {
		mergedList[k] = list[i]
		k++
		i++
	}

	for j <= e {
		mergedList[k] = list[j]
		k++
		j++
	}

	for k = 0; k < len(mergedList); k++ {
		list[s+k] = mergedList[k]
	}
}

func (m *mergeSorterInversionCount) mergeSortInversionCount(list []int, s, e int) {
	if s < e {
		mid := s + (e-s)/2
		m.mergeSortInversionCount(list, s, mid)
		m.mergeSortInversionCount(list, mid+1, e)
		m.mergeInversionCount(list, s, mid, e)
	}
}

func inversionCount() {
	var t, n int
	scanf("%d\n", &t)
	m := mergeSorterInversionCount{}
	for t > 0 {
		scanf("\n")
		scanf("%d\n", &n)
		list := make([]int, n)
		for i := 0; i < n; i++ {
			scanf("%d\n", &list[i])
		}
		m.invCount = 0
		m.mergeSortInversionCount(list, 0, n-1)
		printf("%d\n", m.invCount)
		t--
	}
}
